### DISCUSSION TITLE
Built-in web server shows logging output - strange?

### DISCUSSION THREAD

**Discussion by glow12121** [datenstrom/community#536] [Ask a question] [answered]
Hi,

I recently used Yellow after a several months pause and started to get these strange errors when `serving` the website. I tried to download the latest Yellow `master` and some extensions, and even rolled back to the default Berlin theme, with only minor changes to CSS and layout. But still, the errors are there. When `served`, I don't notice any problems with the actual webpages; nor are there any problems when the site is `built`.

```
[Fri Aug 14 03:03:50 2020] 127.0.0.1:56316 [200]: /
[Fri Aug 14 03:03:50 2020] 127.0.0.1:56320 [200]: /media/themes/berlin.png
[Fri Aug 14 03:03:50 2020] 127.0.0.1:56324 [200]: /media/extensions/bundle-89ba1947ac.min.css
[Fri Aug 14 03:03:50 2020] 127.0.0.1:56328 [200]: /media/extensions/bundle-ae5793b138.min.js
[Fri Aug 14 03:03:51 2020] 127.0.0.1:56332 [200]: /media/images/sunrise.jpg
[Fri Aug 14 03:03:51 2020] 127.0.0.1:56336 [200]: /media/extensions/fontawesome.woff
[Fri Aug 14 03:03:51 2020] 127.0.0.1:56340 [200]: /media/themes/berlin.png
```

**Comment by markseuffert** [comment]
Which PHP version and operating system do you have? Did you recently update your PHP or operating system? When is this output shown, which looks like a web server log by the way, is it shown always or after some time? Do you have additional information for us?

**Comment by glow12121** [comment]
Thank you for replying, Mark. I've just tested my Yellow again, the log is below. After your remark, I indeed noticed that the "200" notifications appear each time I open a page on localhost:8000 served site. Could it be not an error but just some kind of browser reports?

I am using Solus OS (Linux), which I regularly update. And PHP today is *[corrected] 7.3.19-80*.

```
user@x240 ~/Mega/Stuff $ cd /home/user/Mega/Stuff/Yellow/yellow-master
user@x240 ~/Mega/Stuff/Yellow/yellow-master $ php yellow.php
Datenstrom Yellow is for people who make small websites.
Syntax: php yellow.php about
        php yellow.php build [directory location]
        php yellow.php check [directory location]
        php yellow.php clean [directory location]
        php yellow.php install [extension]
        php yellow.php serve [directory url]
        php yellow.php uninstall [extension]
        php yellow.php update [extension]
        php yellow.php user [option email password name]
user@x240 ~/Mega/Stuff/Yellow/yellow-master $ php yellow.php serve
Starting built-in web server on http://localhost:8000/
Press Ctrl-C to quit...
[Mon Aug 31 11:52:10 2020] 127.0.0.1:39524 [200]: /
[Mon Aug 31 11:52:10 2020] 127.0.0.1:39528 [200]: /media/extensions/bundle-89ba1947ac.min.css
[Mon Aug 31 11:52:10 2020] 127.0.0.1:39532 [200]: /media/extensions/bundle-ae5793b138.min.js
[Mon Aug 31 11:52:10 2020] 127.0.0.1:39536 [200]: /media/themes/berlin.png
[Mon Aug 31 11:52:10 2020] 127.0.0.1:39540 [200]: /media/themes/berlin.png
[Mon Aug 31 11:52:10 2020] 127.0.0.1:39544 [200]: /media/images/sunrise.jpg
[Mon Aug 31 11:52:10 2020] 127.0.0.1:39548 [200]: /media/extensions/fontawesome.woff
[Mon Aug 31 11:52:11 2020] 127.0.0.1:39552 [200]: /media/themes/berlin.png
[Mon Aug 31 11:52:25 2020] 127.0.0.1:39570 [200]: /fiction/
[Mon Aug 31 11:52:25 2020] 127.0.0.1:39574 [200]: /media/extensions/bundle-89ba1947ac.min.css
[Mon Aug 31 11:52:25 2020] 127.0.0.1:39578 [200]: /media/extensions/bundle-ae5793b138.min.js
[Mon Aug 31 11:52:25 2020] 127.0.0.1:39582 [200]: /media/themes/berlin.png
[Mon Aug 31 11:52:25 2020] 127.0.0.1:39586 [200]: /media/extensions/fontawesome.woff
^C
user@x240 ~/Mega/Stuff/Yellow/yellow-master $ php yellow.php build
Building static website 100%... done
Yellow build: 194 files, 0 errors
user@x240 ~/Mega/Stuff/Yellow/yellow-master $ 
```

**Comment by markseuffert** [comment]
Thanks for the detailed information. Unfortunately I can't reproduce this on my computer, PHP 7.1.33 on a Macbook. The last thing I see on my computer is `Press Ctrl-C to quit...`,  there are no additional lines when I browse the website. But I remember that I saw something like this 3 years ago and the [PHP documentation mentions it as well](https://www.php.net/manual/en/features.commandline.webserver.php). Each line is a request and for example 200 is the [HTTP status code](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes). 

I am not sure, why the built-in web server behaves different on different computers. 

Anyone else gets logging output after `php yellow.php serve`?

**Comment by glow12121** [comment]
Thank you for these comments. A note on the fields: I am using Vivaldi browser, which is Chromium-based.

```
Vivaldi: 3.2.1967.47 (Stable channel) (64-bit)
User Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.135 Safari/537.36
```

**Comment by schulle4u** [comment]
I've just tested with different php versions and couldn't find any configuration directive which controls internal webserver output logging. However, it seems that this behaviour depends on distribution packaging. On debian 10 the default php version is 7.3, and there is no output logging to stdout, regardless of error reporting settings in php.ini. After updating to php 7.4 from packages.sury.org with default ini settings I'm getting similar log messages to stdout, but only by calling the web server directly: 

```
steffen@steffen-cubi5:~$ cd yellow/
steffen@steffen-cubi5:~/yellow$ php yellow.php serve
Starting built-in web server on http://localhost:8000/
Press Ctrl-C to quit...
[Fri Sep  4 21:04:11 2020] PHP 7.4.9 Development Server (http://localhost:8000) started
^C
steffen@steffen-cubi5:~/yellow$ php -S localhost:8000 yellow.php
[Fri Sep  4 21:05:05 2020] PHP 7.4.9 Development Server (http://localhost:8000) started
[Fri Sep  4 21:05:14 2020] 127.0.0.1:51158 Accepted
[Fri Sep  4 21:05:14 2020] 127.0.0.1:51158 Closing
[Fri Sep  4 21:05:14 2020] 127.0.0.1:51160 Accepted
[Fri Sep  4 21:05:14 2020] 127.0.0.1:51160 Closing
[Fri Sep  4 21:05:15 2020] 127.0.0.1:51161 Accepted
^Csteffen@steffen-cubi5:~/yellow$
```

The output logging only happens to stdout and not in the browser, therefore I don't see a big problem here.

**Comment by markseuffert** [comment]
This morning I had a quick look in the [PHP source code of the web server](https://github.com/php/php-src/blob/master/sapi/cli/php_cli_server.c) and here's what I found out:

The static variable `php_cli_server_log_level` defines how much logging output is generated, default is 3. The function `sapi_cli_server_log_write()` outputs the messages to stderr. There's apparently no setting to control the log level, but there's an undocumented command line argument `q` that decreases the log level. I think this gives us some possibilities to suppress/cleanup the logging output.

Can someone give me access to a computer with logging?

**Comment by markseuffert** [comment]
Thanks @schulle4u for a test environment.

I couldn't find a good solution. It's a strange design choice that the logging output is written to to stderr, because now we can't suppress the logging output without suppressing useful error messages at the same time. I would ignore the output from the built-in web server, until a better solution comes up.

**Comment by markseuffert** [comment, marked as answer]
I tried another solution. The built-in web server is now running in silent mode, which means all logging output will be suppressed unless an error occurs. If the web server can't start, for example when the port number is taken, then one line will be shown that describes what went wrong.

@thumos333 Does this work for you?

**Comment by glow12121** [comment]
> @thumos333 Does this work for you?

Yes, there seems to be no output at all during execution of `serve`. Here is how it looks:

```
~/Mega/Stuff/Yellow/yellow-master $ php yellow.php serve
Starting built-in web server on http://localhost:8000/
Press Ctrl+C to quit...
^C

```

**Comment by markseuffert** [comment]
Lovely, that's what I was hoping to hear. 😁
